### ISSUE TITLE
Getting hashed URLs despite using createBrowserHistory 

### ISSUE THREAD

**Issue by jazlalli** [rackt/react-router#2082]
Hi
I'm upgrading my app from `0.13.3` to `1.0.0-rc1`, as detailed in the upgrade guide, but no matter what I do I'm getting hashed URLs. I've simplified my Router down to just a handful of routes for simplicity, but the problem remains. Router is as below

```javascript
import React from 'react';
import {Router, Route, IndexRoute} from 'react-router';
import createBrowserHistory from 'history/lib/createBrowserHistory';

const router = (
  <Router history={createBrowserHistory()}>
    <Route path="/" component={App}>
      <IndexRoute component={Login} />
      <Route path="oauthcallback" component={OAuth} />
      <Route path="home" component={Home} />

      <Route path="sqms">
        <IndexRoute component={Sqms} />
      </Route>

      <Route path="est">
        <IndexRoute component={Est} />
      </Route>
    </Route>
  </Router>
);

React.render(router, document.getElementById('content'));
```

`/` is all fine. It kicks off an OAuth flow by redirecting to an external URL, which calls back to the `/oauthcallback` route. The OAuth component processes the parameters, and when it's done calls `this.history.pushState(null, '/home');`. I'm using the History mixin for this. That results in the URL path being of the form `/home#/?_k=pc8hae`.

Even if I attempt to remove the hash and reload `/home`, the hash ends up back on the URL. The Home component is a dumb component with nothing other than rendering going on.

As well as the hash on `/home`, the console also shows an error: `Warning: Location "/" did not match any routes`, which doesn't make any sense as `/` renders fine.

Is there anything in what I've described that sounds like a potential cause? Any direction would be helpful as I've been trying to resolve this for the last few hours and am all outta ideas.

Thanks.

**Comment by malte-wessel** [comment]
maybe like this?
```javascript
import { createBrowserHistory } from 'history';
````

**Comment by taurose** [comment]
Browser history is exported as `createHistory` in the entry file so it should be

```js
import {createHistory} from 'history';
```

The original code should work as far as I can tell. I think the error is somewhere else. I'd try to track why and where [createHashHistory](https://github.com/rackt/history/blob/master/modules/createHashHistory.js#L40) is called, since that shouldn't happen with your setup. [This line](https://github.com/rackt/react-router/blob/v1.0.0-rc1/modules/Router.js#L49) is the only place I can think of, but since you pass `history`, it shouldn't come from there.

**Comment by olegakbarov** [comment]
Got exactly same problem. Really appreciate some help!

**Comment by jazlalli** [comment]
I think I've found the cause here, and uncovered a bug in the process.

The cause of this error was my own doing. My routes make use of `IndexRoute` on a couple of occasions, so when I want to hit those routes, I need to use `IndexLink`, as per the [upgrade guide](https://github.com/rackt/react-router/blob/master/UPGRADE_GUIDE.md#linking-to-index-routes).

When doing this, I was getting the following output in the console

```
Warning: React.createElement: type should not be null or undefined. It should be a string (for DOM elements) or a ReactClass (for composite components).
Warning: React.createElement: type should not be null or undefined. It should be a string (for DOM elements) or a ReactClass (for composite components).
Warning: Only functions or strings can be mounted as React components.
Uncaught TypeError: Cannot read property 'toUpperCase' of undefined
```

As part of trying to get around this, I had ended up changing my `IndexLink` imports from

```javascript
import {IndexLink} from 'react-router';
```
to
```javascript
import IndexLink from 'react-router';
```

The `Home` component was trying to render 2 IndexLinks, and so when the Router was firing up, there was no history and so [this line](https://github.com/rackt/react-router/blob/v1.0.0-rc1/modules/Router.js#L49), as mentioned by @taurose, was falling back to `createHashHistory`.

Reverting the import back to the original resolved the presence of the hash, but uncovered another issue which I think is a bug...

`IndexLink` is not included in the 1.0.0-rc1 release!! As you can see [here](https://github.com/rackt/react-router/blob/v1.0.0-rc1/modules/index.js). When I add the import myself, and rebuild locally, everything works! I can't see an issue for this, so I created a new one for it -> https://github.com/rackt/react-router/issues/2099

This issue can be closed. Cheers.

**Comment by prashanw** [comment]
@jazlalli - can you provide a working example of your solution? I'm also trying to implement a very similar oauth solution. Thanks.
